### PR TITLE
Add ‘scale’ for display of a RubParagraph using a RubDisplayScanner

### PR DESCRIPTION
**Pull request by Rinzwind**
Pull request #14630 added ‘scale’ for display of a Paragraph using a BitBltDisplayScanner, this pull request extends that to the display of a RubParagraph using a RubDisplayScanner. An example is given by the snippet below. For comparison, for ‘form1’, the example Rubric paragraph is displayed normally and then the form is scaled by a factor of two:

<p align="center"></p>

For ‘form2’, the paragraph is again displayed directly on the larger form so that the character glyphs have better detail:

<p align="center"></p>

```smalltalk
text := Text streamContents: [ :stream |
	stream
		nextPutAll: (Microdown asRichText: '# Lorem Ipsum'); cr;
		nextPutAll: (Microdown asRichText: 'Dolor **sit** amet, _consectetur adipiscing_ elit,');
		nextPutAll: ' ';
		nextPutAll: (Microdown asRichText: 'sed do eiusmod tempor ![](pharo:///Object/iconNamed:/home) incididunt ut');
		nextPutAll: '  ';
		withAttribute: (TextAnchor new anchoredMorph: (CircleMorph new extent: 10@10; yourself); yourself) do: [
			stream nextPut: Character home ];
		nextPutAll: '  ';
		nextPutAll: (Microdown asRichText: '  labore ~et dolore magna~ aliqua.') ].
(editingArea := RubEditingArea new)
	extent: 300@0;
	beWrapped;
	setTextWith: text.
paragraph := editingArea paragraph next next next.
bounds := 0@0 corner: paragraph extent.
scale := 2.

(baseForm := Form extent: paragraph extent depth: Display depth) fillColor: Color white.
baseForm getCanvas
	rubParagraph: paragraph bounds: bounds color: Color black;
	translateBy: bounds topLeft during: [ :translatedCanvas |
		paragraph text embeddedMorphs do: [ :morph |
			translatedCanvas fullDrawMorph: morph ] ].
form1 := baseForm scaledToSize: baseForm extent * scale.

(form2 := Form extent: form1 extent depth: Display depth) fillColor: Color white.
canvas := form2 getCanvas.
canvas setPaintColor: Color black.
port := canvas privatePort.
origin := canvas origin.
scanner := (port clippedBy: ((bounds translateBy: origin) scaleBy: scale)) rubDisplayScannerFor: paragraph
	foreground: Color black.
scanner scale: scale.
paragraph drawOn: (canvas copyClipRect: (bounds scaleBy: scale)) using: scanner at: (origin + bounds topLeft) * scale.
canvas transformBy: (MatrixTransform2x3 withScale: scale) clippingTo: canvas clipRect during: [ :scaledCanvas |
	scaledCanvas translateBy: bounds topLeft during: [ :translatedCanvas |
		paragraph text embeddedMorphs do: [ :morph |
			translatedCanvas fullDrawMorph: morph ] ] ].

{ form1. form2 } keysAndValuesDo: [ :index :form |
	PNGReadWriter putForm: form onFileNamed: FileLocator imageDirectory / (index asString , '.png') ]
```

### CODE REVIEW

**Comment by Rinzwind** [comment]
I also applied this to the [experiment from the earlier pull requests](https://github.com/pharo-project/pharo/pull/14630#issuecomment-1712915201): commits dc4feb418969674d, f7849fdc6a8ce119 and 853be80e38cab1dc (these commits are not included in this pull request). In the extra world opened by `#example` on OSWorldRendererScalingCanvas, the font scaling is now finally also used for methods in a System Browser:

<p align="center">3B</p>

I’m not sure that the changes to FormCanvas make sense, these expose the ‘scale’ parameter in a few of its drawing methods for the sake of a ScalingCanvas wrapped around a FormCanvas, it might be better to turn ScalingCanvas into a subclass of FormCanvas.

File: src/Rubric/RubCharacterScanner.class.st [hunk 1]
```diff
@@ -85,7 +85,6 @@ RubCharacterScanner >> basicScanCharactersFrom: startIndex to: stopIndex in: sou
 	reached, then return stops at: 257. Optional.
 	See Object documentation whatIsAPrimitive."
 	| ascii nextDestX char floatDestX widthAndKernedWidth nextChar atEndOfRun |
-	<primitive: 103>
```
Comment (jecisc) [inline src/Rubric/RubCharacterScanner.class.st:88]: Is it normal to remove the call to a primitive and only keep the fallback code?

Comment (Rinzwind) [reply src/Rubric/RubCharacterScanner.class.st:88]: See the ‘pharo-vm’ commit referred to in commit bb13a55ff2ba9c52: https://github.com/pharo-project/pharo-vm/commit/b80f660c5c9b1bbba5bc. In the primitive table, 103 was changed to `#primitiveFail`.

Comment (Ducasse) [reply src/Rubric/RubCharacterScanner.class.st:88]: Hi kris all your work is super nice :)
I hope that you will come to the sprint at the end of the month. 
Do you have a code snippet to enable this nice looking changes on our poor Retina screens :)
I know it should be somewhere but I do not remember.

**Comment by Ducasse** [comment]
And yes in future iteration we may introduce a subclass for the canvas

**Review by jecisc — APPROVED** [review]

File: src/Rubric/RubCharacterScanner.class.st (hunk 1 above)
Comment (Rinzwind) [reply src/Rubric/RubCharacterScanner.class.st:88]: Thanks! I will put the snippet in a separate comment below.

**Comment by Rinzwind** [comment]
@Ducasse, for the ‘Retina display’ scaling, there are now two different approaches you can try:

* The approach using the display scale factor: in Pharo 12 or 11, do this snippet (from [pull request #13848](https://github.com/pharo-project/pharo/pull/13848)).
  ```smalltalk
  WorldMorph displayScaleFactor: 2.
  SourceCodeFonts setSourceCodeFonts: SourceCodeFonts sizeSmall * WorldMorph displayScaleFactor.
  ToggleWithSymbolMenuItemShortcut
  	classVarNamed: 'SymbolFont'
  		put: (LogicalFont familyName: 'Lucida Grande' pointSize: 10 * WorldMorph displayScaleFactor);
  	classVarNamed: 'SymbolTable' put: nil.
  ThemeIcons current: ((ThemeIcons named: 'svgPack') in: [ :themeIcons |
  	(themeIcons respondsTo: #loadIconsFromUrlUsingScale:) ifTrue: [
  		themeIcons loadIconsFromUrlUsingScale: WorldMorph displayScaleFactor ] ];
  	yourself).
  World scaleFactor: WorldMorph displayScaleFactor reciprocal.
  SpFontStyle applyDisplayScaleFactor: false.
  ```
* The approach using ScalingCanvas: create an image in Pharo Launcher using Pharo 12 build 751 as the template, repair the repository ‘pharo’ by cloning from GitHub using ‘Rinzwind’ as the ‘Owner name’ and ‘pharo’ as the ‘Project name’, check out the branch ‘scaling-canvas-tryout-4’ from ‘origin’, then do `OSWorldRendererScalingCanvas example` to open an extra world that is drawn using ScalingCanvas. Note that context menus may open in the main world ([issue #14585](https://github.com/pharo-project/pharo/issues/14585)).

The problem with the first approach is the need to take the display scale factor into account everywhere (as in pull requests [#1461](https://github.com/pharo-project/pharo/pull/1461), [#1719](https://github.com/pharo-project/pharo/pull/1719), [#1852](https://github.com/pharo-project/pharo/pull/1852), [#13848](https://github.com/pharo-project/pharo/pull/13848), [#13850](https://github.com/pharo-project/pharo/pull/13850), [#13953](https://github.com/pharo-project/pharo/pull/13953) and so on). That is no longer necessary in the second approach, as the scaling is handled by the canvas.

**Comment by Rinzwind** [comment]
I’ve made some improvements to [the experiment](https://github.com/pharo-project/pharo/pull/14665#issuecomment-1720133696): commits d5726f900d794519, 401cc456c8f7d986 and bcc1c13c078f19cd. To try these: follow the [earlier instructions](https://github.com/pharo-project/pharo/pull/14665#issuecomment-1721624999) for the second approach, but use Pharo 12 build 820 as the template and check out ‘scaling-canvas-tryout-5’, then do `OSWorldRendererScalingCanvas example`. Keep in mind: as previously noted, context menus may open in the main world ([issue #14585](https://github.com/pharo-project/pharo/issues/14585)); another issue affecting extra worlds is that the resize cursor isn’t shown ([issue #14817](https://github.com/pharo-project/pharo/issues/14817)).

**Pull request merged**